File: src/main/java/coza/trojanc/receipt/loader/XmlLoader.java
package coza.trojanc.receipt.loader;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.xml.XmlMapper;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.lang.reflect.ParameterizedType;

/**
 * @author Charl Thiem
 */
public class XmlLoader<T> extends AbstractLoader<T> {

	protected static final ObjectMapper MAPPER = new XmlMapper();

	protected Class<T> clazz;

	public XmlLoader(){
		this.clazz = (Class) ((ParameterizedType) getClass().getGenericSuperclass()).getActualTypeArguments()[0];
	}

	@Override
	public T load(InputStream inputStream) throws IOException {
		return load(inputStream, MAPPER, clazz);
	}

	@Override
	public T load(String jsonString) throws IOException {
		return load(jsonString, MAPPER, clazz);
	}

	@Override
	public void write(T instance, OutputStream out) throws IOException {
		write(instance, out, MAPPER);
	}
}

File: receipt-java/src/main/kotlin/coza/trojanc/receipt/loader/XmlLoader.kt
package coza.trojanc.receipt.loader

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.xml.XmlMapper

import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.lang.reflect.ParameterizedType

/**
 * @author Charl Thiem
 */
abstract class XmlLoader<T> : AbstractLoader<T>(XmlMapper()) {

    protected var clazz: Class<T>

    init {
        this.clazz = (javaClass.genericSuperclass as ParameterizedType).actualTypeArguments[0] as Class<T>
    }

    @Throws(IOException::class)
    override fun load(inputStream: InputStream): T {
        return load(inputStream, getMapper(), clazz)
    }

    @Throws(IOException::class)
    override fun load(jsonString: String): T {
        return load(jsonString, getMapper(), clazz)
    }

    @Throws(IOException::class)
    override fun write(instance: T, out: OutputStream) {
        write(instance, out, getMapper())
    }

}
